docs: migrate documentation from Gatsby to MkDocs material

am29d opened the pull request:
**Issue #, if available:**

## Description of changes:

Due to a rather high time investment for documentation maintenance (read: fight npm ecosystem), we decided to move the documentation to [MkDocs material](https://squidfunk.github.io/mkdocs-material/). This pull contains the migration of the content to the MkDocs. 

Here is the live preview: https://am29d.github.io/aws-lambda-powertools-python/

The github workflow `python_docs.yml` contains for now the branch `docs/mkdocs` to have a live preview in my fork. It will be removed once the PR is approved. 

**Checklist**

<!--- Leave unchecked if your change doesn't seem to apply --> 

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [ ] Update tests
* [x] Update docs
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/376ec0a2ac0d2a40e0af5717bef42ff84ca0d1b9/.github/semantic.yml#L2)

## Breaking change checklist

<!--- Ignore if it's not a breaking change -->

**RFC issue #**:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

cakepietoast commented on docs/utilities/validation.md:
```diff
@@ -22,9 +20,8 @@ You can also use the standalone `validate` function, if you want more control ov
-    a <strong>dictionary</strong>, not a filename.
+!!! warning
+    Both `validator` decorator and `validate` standalone function expects your JSON Schema to be a **dictionary**, not a filename.
 </Note>
```
```suggestion

```

cakepietoast commented on docs/utilities/validation.md:
````diff
@@ -34,16 +31,18 @@ We support any JSONSchema draft supported by [fastjsonschema](https://horejsek.g
+        return event
+    ```
 
 **NOTE**: It's not a requirement to validate both inbound and outbound schemas - You can either use one, or both.
````
```suggestion
!!! note
    It's not a requirement to validate both inbound and outbound schemas - You can either use one, or both.
```

cakepietoast reviewed (changes requested):
Great change! Added some comments inline. Not sure if its just me but the content seems quite squashed in the center of the page, so you need to do a bit of side scrolling for some of the code blocks. Can we add a stylesheet to increase the width of `.md-grid`?

heitorlessa commented:
> Great changes! Added some comments inline. Not sure if its just me but the content seems quite squashed in the center of the page, so you need to do a bit of side scrolling for some of the code blocks. Can we add a stylesheet to increase the width of `.md-grid`?

Same here - Experiment using a `max-width` between 70rem to 77rem to see how that would look like. [You will need an additional CSS](https://squidfunk.github.io/mkdocs-material/customization/#additional-css) and use `!important` to override the theme.

```css
.md-grid
    max-width: 75rem
```

I've committed most of @cakepietoast suggestions, so we can spend more time on the larger pieces of content we'll also need to change to best use the content area we have now :) 

If you're not comfortable with the CSS pieces, leave that with me - I can also adjust the `blockquotes` I see on Tom's feedback too.

am29d commented on docs/utilities/parameters.md:
````diff
@@ -78,46 +81,52 @@ The AWS Systems Manager Parameter Store provider supports two additional argumen
+        value = secrets_provider.get("my-secret")
+    ```
 
 ## DynamoDB
````
Gotcha, looks indeed a little bit odd.

am29d commented:
> > Great changes! Added some comments inline. Not sure if its just me but the content seems quite squashed in the center of the page, so you need to do a bit of side scrolling for some of the code blocks. Can we add a stylesheet to increase the width of `.md-grid`?
> 
> Same here - Experiment using a `max-width` between 70rem to 77rem to see how that would look like. [You will need an additional CSS](https://squidfunk.github.io/mkdocs-material/customization/#additional-css) and use `!important` to override the theme.
> 
> ```css
> .md-grid
>     max-width: 75rem
> ```
> 
> I've committed most of @cakepietoast suggestions, so we can spend more time on the larger pieces of content we'll also need to change to best use the content area we have now :)
> 
> If you're not comfortable with the CSS pieces, leave that with me - I can also adjust the `blockquotes` I see on Tom's feedback too.

removed blockquotes and increased `max-width` to 75rem, looks much better now.

am29d commented on docs/core/tracer.md:
```diff
@@ -170,42 +162,48 @@ def get_s3_object(bucket_name, object_key):
+!!! warning
+    **We do not support async Lambda handler** - Lambda handler itself must be synchronous
 
 You can trace asynchronous functions and generator functions (including context managers) using `capture_method`.
```
The change was proposed by @heitorlessa , any thoughts on that?

heitorlessa replied:
I'll look into it this week more carefully.

The content tab suggestion was to save reader screen, so one can look at what's relevant instead of every option at once.

If I understood Tom's note is around the code split lacking the Lambda handler made it more confusing, and instead have a long block with all options (as it was)

cakepietoast replied:
We could still split it into 3 separate blocks if we want to save screen space, but in that case we should copy the lambda handler code into each block. In that case, the lambda handler code should only show it calling the relevant function (either async, generator or context mgr rather than all 3).

heitorlessa commented:
I've made a mistake on the units (75rem), viewport width is better here - Could you update to `max-width: 90vw`, please?

Here's before (75rem) - It looks great in wide screens but looks cramped on laptop's screen (16inch and smaller)

![image](https://user-images.githubusercontent.com/3340292/106779659-7feece00-6647-11eb-811b-29934f47af73.png)

Here's after (90% of viewport width):

![image](https://user-images.githubusercontent.com/3340292/106779811-a57bd780-6647-11eb-8ccb-64839f638ab1.png)

heitorlessa commented on docs/stylesheets/extra.css:
```diff
@@ -0,0 +1,3 @@
+.md-grid {
+    max-width: 75rem
+}
```
Another is for text in Tables - Font size is reduced and harder to read. here's a fix (try that first, just in case):

```css
.md-typeset table:not([class]) {
    font-size: 0.75rem
}
```

This should make it more accessible for those with low eyesight - Feel free to experiment with other sizes too

heitorlessa replied:
This will also enhance the visual aid for accessibility purposes (screen readers), and those with low eyesight or colorblind by adding a visual cue it's a link with a tiny bottom border

```css
.md-typeset a {
    border-bottom: 0.1px dashed black
}
```

heitorlessa replied:
Quick fix to increase emphasis on active navigation anchors as you scroll the page, and the menu that's being actively viewed - Also for low eyesight

```css
.md-nav__link--active {
    font-weight: bold
}
```

It's subtle, you'll only notice as you scroll the page on your right (right sidemenu)

heitorlessa replied:
Another one to fix low eyesight issues when trying to read non-bold text in warnings, info, boxes, etc.

```css
.md-typeset .admonition, .md-typeset details {
    font-size: 0.70rem
}
```

Maybe needs !important, not sure, worth checking

am29d replied:
added proposed changes

heitorlessa commented on docs/index.md:
````diff
@@ -0,0 +1,167 @@
+
+=== "template.yml"
+
+    ```yaml
````
perhaps for completion (can be done later, don't bother if you're already got a lot on your plate)... add an IAM Role or Policy resource with this policy in it.

am29d replied:
Added an IAM role that will be assumed by CloudFormation.

heitorlessa commented on docs/core/tracer.md:
```diff
@@ -88,80 +86,74 @@ def handler(event, context):
-  <br/><br/>
-</Note><br/>
+!!! warning
+    **Can exceptions contain sensitive information from your Lambda handler or functions, where Tracer is used?**
```
Typo of my own making that we can fix now :D

```suggestion
    **Where Tracer is used, can your exceptions contain any sensitive information from your Lambda handler or functions?**
```

heitorlessa commented on docs/core/tracer.md:
```diff
@@ -88,80 +86,74 @@ def handler(event, context):
-### Annotations
+### Annotations & Metadata
 
 Annotations are key-values indexed by AWS X-Ray on a per trace basis. You can use them to filter traces as well as to create [Trace Groups](https://aws.amazon.com/about-aws/whats-new/2018/11/aws-xray-adds-the-ability-to-group-traces/).
```
```suggestion
**Annotations** are key-values indexed by AWS X-Ray on a per trace basis. You can use them to filter traces as well as to create [Trace Groups](https://aws.amazon.com/about-aws/whats-new/2018/11/aws-xray-adds-the-ability-to-group-traces/).
```

heitorlessa commented on docs/core/tracer.md:
```diff
@@ -88,80 +86,74 @@ def handler(event, context):
-
-### Metadata
-
 Metadata are non-indexed values that can add additional context for an operation.
```
```suggestion
**Metadata** are non-indexed values that can add additional context for an operation.
```

heitorlessa reviewed (changes requested):
Suggestions to improve low eyesight issues, color blind friendly, and increase emphasis on links, active anchors, and text size in informational boxes.

Nr18 commented on Makefile:
```diff
@@ -10,7 +10,7 @@ dev:
 
 dev-docs:
-	cd docs && yarn install
+	pip install "mkdocs-material"
```
We are actually using that docker for our internal documentation. Works perfectly also detects changes 👌

am29d replied:
updated `pyproject.toml` and `Makefile` with new target.

am29d commented:
I have changed the tab size for the `template.yml` examples to 2 to make it more readable, don't see much value to spend lots of whitespace so we can then force horizontal scroll. The python examples remains with 4 spaces, since they are quite compact.

michaelbrewer commented:
@cakepietoast @heitorlessa  - nice work so far, please just check the hyperlinks and the anchor tags work ;)

heitorlessa commented:
We should try "strict: true" in mkdocs config to catch these broken links.

If not, I believe markdownlint can catch them and other syntax errors too

cakepietoast commented on docs/core/tracer.md:
````diff
@@ -0,0 +1,369 @@
+        tracer.put_metadata(key="payment_response", value=ret)
+    ```
+
+## Synchronous functions
````
I think this should be under the getting started header?
```suggestion
### Synchronous functions
```

cakepietoast commented on docs/core/tracer.md:
````diff
@@ -0,0 +1,369 @@
+    return ret
+```
+
+## Asynchronous and generator functions
````
Same as above, I think it should be nested under getting started

```suggestion
### Asynchronous and generator functions
```

heitorlessa commented:
THANK YOU to everyone who helped us in this migration from Gatsby to MKdocs.. in no particular order 

@am29d @michaelbrewer @cakepietoast @Nr18 

This should be out today
